Add: Redis Geospatial Command Stages (#328)
codecrafters-io/build-your-own-redis: stage_descriptions/geospatial-02-ck3.md

In this stage, you'll add support for validating the latitude and longitude provided in the `GEOADD` command.

### The `GEOADD` command (Validate coordinates)
The latitude and longitudes used in the GEOADD command should be in a certain range. Valid longitudes are from -180 to 180 degrees. Valid latitudes are from -85.05112878 to 85.05112878 degrees. Both of these limits are inclusive.
Example use case:

```
# Invalid latitude
> GEOADD places 180 90 test1
(error) ERR invalid longitude,latitude pair 180.000000,90.000000

# Invalid longitude
> GEOADD places 181 0.3 test2
(error) ERR invalid longitude,latitude pair 181.000000,0.300000
```

### Tests
The tester will execute your program like this:
```
./your_program.sh
```
It will then send multiple GEOADD commands specifying one or more location to add. For valid co-ordinates, the tester will expect the response to be the usual response of the GEOADD command. For invalid co-ordinate values, it will expect an error.

For example, the tester will expect the response of the following command

```
$ redis-cli

> > GEOADD location_key 200 100 foo

Reviewer: Let's keep the command and response in the same code block to be consistent, also use `redis-cli GEOADD ...` so it's simpler to read

```

to be
```
(error) ERR invalid longitude,latitude pair 200,100
```

which is RESP-encoded as

```
-ERR invalid longitude,latitude pair 200,100\r\n
```

### Notes

> - In case of invalid co-ordinates, the tester is lenient in checking error messages so you don't have to stick to the exact format Redis uses. The exact format it checks for is `invalid` (case-insensitive). Examples of error message strings that will pass the tests:

Reviewer: Can we check for a bit more than just "invalid"? Maybe check for presence of `latitude` / `longitude` too, depending on which is invalid?

> - In case of invalid co-ordinates, the tester is lenient in checking error messages so you don't have to stick to the exact format Redis uses. The exact format it checks for is `invalid` (case-insensitive). Examples of error message strings that will pass the tests:

Author: I am not sure if I understood correctly. Assuming that we ought to check for which of the values (latitude/longitude) is incorrect and checking for the indication in the error response, Redis does not specify which value `latitude`/`longitude` is invalid.
```
> GEOADD p 15 240 foo
(error) ERR invalid longitude,latitude pair 15.000000,240.000000

> GEOADD p 240 23 foo
(error) ERR invalid longitude,latitude pair 240.000000,23.000000
```

So, I am not sure if we should test this.

We can instead add another test case. For e.g. with a non-numeric values:
```
127.0.0.1:6379> GEOADD p 15 foo bar
(error) ERR value is not a valid float
```

    - `invalid longitude,latitude pair 200,100`
    - `Invalid longitude,latitude`
    - `invalid`

> - If a `GEOADD` command specifies multiple locations, and if the co-ordinates of any one of those locations is invalid, it returns the error message and all the locations are discarded.

Reviewer: We don't cover adding multiple locations so we can remove this

In this stage, you'll add support for validating the latitude and longitude provided in the `GEOADD` command.

### The `GEOADD` command (Validate coordinates)
Latitudes and longitudes used in the `GEOADD` command should be in a certain range as per [EPSG:3857](https://epsg.io/3857). Valid longitudes are from -180 to 180 degrees. Valid latitudes are from -85.05112878 to 85.05112878 degrees. Both of these limits are inclusive.

> Example use case:

Reviewer: > If either of these values aren't within the appropriate range, `GEOADD` returns an error. Examples: 

(We need to explain what we're showing examples of - just "example use case" doesn't fully explain what's happening)

```bash
# Invalid latitude
> GEOADD places 180 90 test1
(error) ERR invalid longitude,latitude pair 180.000000,90.000000

# Invalid longitude
> GEOADD places 181 0.3 test2
(error) ERR invalid longitude,latitude pair 181.000000,0.300000
```

### Tests
The tester will execute your program like this:

```bash
$ ./your_program.sh
```

> It will then send multiple `GEOADD` commands specifying one or more location to add. For valid coordinates, the tester will expect the response to be the usual response of the `GEOADD` command. For invalid coordinates values, it will expect an error.

Reviewer: ```suggestion
It will then send multiple `GEOADD` commands. If the coordinates supplied are valid, the tester will expect the response to be `:1\r\n`, the usual response of the `GEOADD` command. For invalid coordinates values, it will expect an error response.
```

For example, the tester might send your program a command like this:

```bash
$ redis-cli GEOADD location_key 200 100 foo

# Expecting error
(error) ERR invalid longitude,latitude pair 200,100
```

The value is RESP simple error, which is RESP-encoded as

```bash
-ERR invalid longitude,latitude pair 200,100\r\n
```

### Notes

- In this stage, you'll only extend responding to `GEOADD` in case of invalid values of latitude and longitude. You do not need to implement the storage mechanism yet.

> - In case of out-of-bounds values for latitude or longitude, the tester is lenient but structured in checking error messages.

Reviewer: ```suggestion
- The tester is lenient in checking error messages, it doesn't require that errors match Redis's exact format.
```

    - The error response does not need to match Redis’s exact error message format.
    - The error response must:
        - Start with the phrase "ERR invalid" (case-insensitive)
         - Contain the word latitude if the latitude value is invalid
        - Contain the word longitude if the longitude value is invalid
        - Contain both "latitude" and "longitude" if both the latitudes and longitudes are invalid.

>     - Examples that will pass:

Reviewer: This is TOO much explanation for a simple error message check, let's simplify. 

For starters, let's not test the "latitude and longitude are both invalid" case. That way we know to assert for exactly one each time. And here let's just mention example of 2-3 error messages that will pass checks.

Find a way to bring this down to 3-4 bullet points tops, no more
